### DISCUSSION TITLE
STM32L5 Using Mongoose over RNDIS and USB CDC

### DISCUSSION THREAD

**Discussion by sckid1108** [cesanta/mongoose#1922] [Ask questions, get answers] [answered]
Has anyone gotten this to work ?

**Comment by scaprile** [comment]
Mongoose runs either over a Berkeley-compatible socket API or using its own TCP/IP stack with a driver over an Ethernet frame transport (including a device like pcap or tap). Can you provide any one of those ? What is your environment (OS, etc) ? What is your use case ?

**Comment by sckid1108** [comment]
Baremetal STM32L4 or L5 device using USB CDC RNDIS Stack From the HAL
Libraries.
See attached  for the STM32 CDC RNDIS Class Library.


Michael Margolese
Principal Engineer
Tenuvah Designs
http://www.tenuvahdesigns.com
Phone: (424) 288-9585

**Reply by scaprile** [reply]
Sorry, attachments on replies don't seem to go through, you have to manually add them.
As far as I get, the CDC class will provide a modem interface and RNDIS will provide an Ethernet-like interface on top of that. STM32CubeIDE doesn't seem to integrate lwIP on that, if you manually integrate it you could use it, perhaps writing an interface (netif). Or, you might try our own stack.
I forwarded this information to our Support Department.

**Reply by cpq** [reply]
Thanks @scaprile ! @sckid1108 , it looks like it is quite possible to do - could you put together a brief description of what you're trying to achieve on a high level, please? Feel free to email support@cesanta.com to setup a quick call to discuss that.

**Comment by sckid1108** [comment]
I want to do Ethernet over USB and use LWiP as the networking layer.
Simple concept is that the embedded device can interact with a browser
using WebSockets
for being able to control devices from webpages or over the internet.

Here is an example project that I have mostly integrated but I am no expert
in LWiP
Device enumerates as a CDC ACM device (Virtual Com Port) *but* does not get
a proper driver from windows to show up as an ethernet device
Dont know if i am using the LWiP stack incorrectly.
https://www.dropbox.com/t/0hPg5khaUG1gnZ3t
Let me know what you think

Michael Margolese
Principal Engineer
Tenuvah Designs
http://www.tenuvahdesigns.com
Phone: (424) 288-9585

**Reply by scaprile** [reply]
This is baremetal, lwIP works with NO_SYS=1 hence it does not provide a socket-compatible interface, so Mongoose would need an adaptation layer to work with lwIP's RAW API.

**Comment by sckid1108** [comment]
Hey Guys,
Here is a much better example I got working on my STM32L422KBT board.
https://www.dropbox.com/t/nmUJp3phrQUdkyu8

Uses TinyUSB with LWiP to serve a web page etc.  Would be SOOO much better
with Mongoose.

Michael Margolese
Principal Engineer
Tenuvah Designs
http://www.tenuvahdesigns.com
Phone: (424) 288-9585

**Reply by cpq** [reply]
@sckid1108 we've added a mongoose example for tinyusb + rndis, see https://github.com/cesanta/mongoose/tree/master/examples/stm32/nucleo-f429zi-usb-rndis

That is a bare metal example, which you can apply with almost no changes to other STM MCUs.
All you need to do, is to setup USB pins and enable USB FS.

Clone mongoose repo, go to this directory and type `make`. You need ARM GCC installed.
That was tested on F429 board, on MacOS:
<img width="1232" alt="Screenshot 2022-12-19 at 14 37 28" src="https://user-images.githubusercontent.com/601816/208450122-c7227a82-66a2-43c0-9374-94d4b72a924e.png">

Talking to a device web server from the workstation:

```
$ curl 192.168.3.1
hi
```

**Comment by sckid1108** [comment]
AMAZING !!! I will port to my board and test as well !

Michael Margolese
Principal Engineer
Tenuvah Designs
http://www.tenuvahdesigns.com
Phone: (424) 288-9585

**Reply by cpq** [reply]
@sckid1108 we might prep the STM32L example as well.
Out of curiosity, what middleware do you use ? FreeRTOS ?

**Comment by sckid1108** [comment]
I will send you my code if i get it working.

For shipping products we ship BAREMETAL.  No RTOS at all.

Michael Margolese
Principal Engineer
Tenuvah Designs
http://www.tenuvahdesigns.com
Phone: (424) 288-9585

**Comment by sckid1108** [comment]
STM32 HAL Libraries also

Michael Margolese
Principal Engineer
Tenuvah Designs
http://www.tenuvahdesigns.com
Phone: (424) 288-9585

**Reply by cpq** [reply]
@sckid1108 baremetal, awesome.
That's what our example is. We don't even use HAL, only CMSIS headers.

**Comment by sckid1108** [comment, marked as answer]
Hey Gang,

I got it working.  (Didn't need netif/ethernet.h since ETHERNET header
length already defined in TUSB).
I assume you guys added DHCP server etc ?  (didn't see this in mongoose i
originally downloaded, thats why i couldnt get it working)

BTW mongoose is now totally *bad ass* and I will be using this instead of
LwIP for our projects in the future.

Tested on Windows 10 (automagically uses the built-in driver. Attached is
my complete build directory.
I wouldn't mind doing a STM32L5 (Cortex-M33 version).

[image: image.png]


[image: image.png]

*MEMORY USAGE USING HAL AND USART FOR DEBUG (Debug code build)*
[image: image.png]
*MEMORY USAGE FOR RELEASE*
[image: image.png]

mongoose_custom.h

#pragma once

#include <errno.h>

#define MG_ARCH MG_ARCH_NEWLIB
#define MG_ENABLE_MIP 1
#define MG_ENABLE_CUSTOM_MILLIS 1
#define MG_IO_SIZE 256

Let me know if you want anything else from me.   BTW We use HAL often to
get things working during development and
to do things like configuration for clocks, IO, etc.

Michael Margolese
Principal Engineer
Tenuvah Designs
http://www.tenuvahdesigns.com
Phone: (424) 288-9585

**Comment by cpq** [comment]
@sckid1108 awesome, thanks for letting us know!

- *DCHP server* - yes, the recent change that added that rndis example, also added a simple DHCP server to the network stack.
- `netif/ethernet.h` - you mention that file is already present: yes and no! it is present if you use LWIP with tinyusb. That file is part of LWIP, basically. If you don't use lwip, then it is not present, that's why we need to add it. I've already filed a PR to tinyusb to get rid of that include: https://github.com/hathach/tinyusb/pull/1800

Would you be able to share some info about the end product, and specifically - that USB use case, please? The email is support@cesanta.com if you do not want it open to the public.

**Comment by groutoutlook** [comment]
hi, great work you guys... but I hope someone can shed some lights into it as well. I'm starting my journey on webusn here as well, trying to utilize USB CDC for interfacing with LAN work... any section of code I should take a look there?

**Reply by sckid1108** [reply]
Apologies but I have not looked at this codebase for ~4 years and mongoose has updated quite a bit. Are you using STM32 by any chance ?

**Reply by sckid1108** [reply]
[stm32_mongoose_usb_gadget.pdf](https://github.com/user-attachments/files/26916938/stm32_mongoose_usb_gadget.pdf)

**Reply by groutoutlook** [reply]
I'm using AT32F435 chip, quite a niche one, but thanks for the pointer.

**Reply by sckid1108** [reply]
OK Mongoose should work independent of which device you are using. You will need to adapt TinyUSB or use the USB Device library supplied with AT32F435 from ArteryTek

**Reply by sckid1108** [reply]
Looks like ArteryTek has at least a CDC USB device class
https://github.com/ArteryTek/AT32F435_437_Firmware_Library/tree/master/middlewares/usbd_class and winusb

**Comment by scaprile** [comment]
@groutoutlook Please no necroposting. We do have working RNDIS examples in the repo, what _exactly_ is your question/problem ? AFAICS, Michael is right at the spot. Please open your own discussion for that, if you need.
@sckid1108 Thanks for dropping by, Michael !

**Reply by groutoutlook** [reply]
Firstly, sorry for that, I just need some help at that moment.
Secondly, you are a new level of mean aren't you? Please delete my response or mark it off-topic if you could.
